pkg/kubelet/kubelet_node_status.go

```diff
@@ -593,6 +593,19 @@ func (kl *Kubelet) tryUpdateNodeStatus(ctx context.Context, tryNumber int) error
 	return err
 }
 
+func (kl *Kubelet) addRandomIntervalToNodeStatusReportFrequency(changed bool) (shouldPatchNodeStatus bool) {
```

Comment:
it is a bit overloaded for my taste, but we were back and forth so I don't want to stop progress, vur I think it will be more clear if you describe with comments all the steps and encapsulate only the logic to "isUpdateStatusPeriodExperid() bool" and "calculateDelay() time.Duration", so both are testable

if I understand now correctly the logic this should be something like
new func
```go
func (kl *kubelet) isUpdateStatusPeriodExperid bool {
if kl.lastStatusReportTime.IsZero() {
     return false
}
if kl.clock.Since(kl.lastStatusReportTime) >= (kl.nodeStatusReportFrequency+kl.delayAfterNodeStatusChange) {
   return true
}
return false
}
```
```go
// no need to update the status yet
if !changed && !kl.isUpdateStatusPeriodExperid() {
	kl.markVolumesFromNode(node)
        return nil
}
// We need to update the node status, if this is caused by a node change we want to calculate a new
// random delay so we avoid all the nodes to reach the apiserver at the same time. If the update is not related
// to a node change, because we run over the period, we reset the random delay so the node keeps updating
// its status at the same cadence
if changed {
     kl.delayAfterNodeStatusChange = kl.calculateDelay()
 else if {
     kl.delayAfterNodeStatusChange = 0
} 
updatedNode, err := kl.patchNodeStatus(originalNode, node)
 if err == nil {
		kl.markVolumesFromNode(updatedNode)
}

```

Reply:
Thanks for your suggestion!
I agree this looks better. I will make the change shortly.